crates/bevy_ecs/src/system/query.rs

```diff
@@ -11,102 +11,171 @@ use bevy_tasks::TaskPool;
-/// filter out entities that do not meet the requirement set by the predicate. [`With`](crate::query::With)
-/// is one such filter and all it does is filter out all entities that do not contain the component
-/// it requests. Let's look at an example on how to use this filter.
+/// Inserting [`Entity`](crate::entity::Entity) in the type parameter tuple will give access
+/// to the entity handle.
 ///
 /// ```
+/// # use bevy_ecs::entity::Entity;
 /// # use bevy_ecs::system::IntoSystem;
 /// # use bevy_ecs::system::Query;
+/// # struct ComponentA;
+/// # struct ComponentB;
+/// # fn system(
+/// query: Query<(Entity, &ComponentA, &ComponentB)>
+/// # ) {}
+/// # system.system();
+/// ```
+///
+/// ## Query filtering
+///
+/// From the second type parameter on, filters can be added to restrict the query results,
+/// filtering out the elements that don't satisfy the given condition.
+///
+/// ```
 /// # use bevy_ecs::query::With;
-/// struct Person(String);
-/// struct IsTallEnough;
-/// fn system(query: Query<&Person, With<IsTallEnough>>) {
-///     for person in query.iter() {
-///         println!("{} is tall enough!", person.0);
-///     }
-/// }
+/// # use bevy_ecs::system::IntoSystem;
+/// # use bevy_ecs::system::Query;
+/// # struct ComponentA;
+/// # struct ComponentB;
+/// # struct ComponentC;
+/// # fn system(
+/// // `ComponentC` data won't be accessed, but only entities that contains it will be queried.
+/// query: Query<(&ComponentA, &ComponentB), With<ComponentC>>
+/// # ) {}
+/// # system.system();
+/// ```
+///
+/// See the [`query`](crate::query) module for a full list of available filters.
+///
+/// ## Optional component access
+///
+/// A component can be made optional in a query by wrapping it into an [`Option`]. In the
+/// following example, the query will iterate over components of both entities that contain
+/// `ComponentA` and `ComponentB` and entities that contain `ComponentA` but not `ComponentB`.
+///
+/// ```
+/// # use bevy_ecs::system::IntoSystem;
+/// # use bevy_ecs::system::Query;
+/// # struct ComponentA;
+/// # struct ComponentB;
+/// # fn system(
+/// query: Query<(&ComponentA, Option<&ComponentB>)>
+/// # ) {}
 /// # system.system();
 /// ```
 ///
-/// As shown above, the filter is a second type parameter of the query. It is optional (defaults to
-/// ()). Filters do not give access to the component data, only limit the entities that the query will match.
+/// If an entity does not contain a component, its corresponding query result value will be
+/// `None`. Optional components increase the number of entities a query has to match against,
+/// therefore they can hurt iteration performance, especially in the worst case scenario where
+/// the query solely consists of only optional components, where all entities will be iterated
+/// over.
+///
+/// ## Single component access
 ///
-/// ### Optional Components
+/// The use of the tuple as the type parameter of `Query` can be omitted if just a single
```

Comment:
IMO the sentence structure here should be inverted for clarity.

"If just a single component needs to be accessed (or filtered by)..."

Reply:
Done. I also added an example in the query filtering section showing usage of multiple filters.